63.CUDAInterop/main.cpp

```diff
@@ -0,0 +1,411 @@
+			params.externalHandleTypes = CCUDADevice::EXTERNAL_MEMORY_HANDLE_TYPE;
+			params.tiling = IGPUImage::TILING::LINEAR;
+			importedimg = mem[2]->exportAsImage(logicalDevice.get(), std::move(params));
+			assert(importedimg);
+		}
+		
+		commandPool = logicalDevice->createCommandPool(queue->getFamilyIndex(), {});
+		bool re = commandPool->createCommandBuffers(IGPUCommandPool::BUFFER_LEVEL::PRIMARY, 1, &cmd, smart_refctd_ptr(m_logger));
+		assert(re);
+
+		auto createStaging = [logicalDevice=logicalDevice]()
+		{
+			auto buf = logicalDevice->createBuffer({ {.size = size, .usage = asset::IBuffer::EUF_TRANSFER_DST_BIT} });
+			auto req = buf->getMemoryReqs();
+			req.memoryTypeBits &= logicalDevice->getPhysicalDevice()->getDownStreamingMemoryTypeBits();
+			auto allocation = logicalDevice->allocate(req, buf.get());
+			assert(allocation.isValid() && buf->getBoundMemory().memory->isMappable());
+			
+			bool re = allocation.memory->map(IDeviceMemoryAllocation::MemoryRange(0, req.size), IDeviceMemoryAllocation::EMCAF_READ);
+			assert(re && allocation.memory->getMappedPointer());
+			memset(allocation.memory->getMappedPointer(), 0, req.size);
+			return buf;
+		};
+
+		stagingbuf = createStaging();
+		stagingbuf2 = createStaging();
+	}
+
+	void launchKernel(CUfunction kernel, CUstream stream)
+	{
+		auto& cu = cudaHandler->getCUDAFunctionTable();
+		// Launch kernel
+		{
+			CUdeviceptr ptrs[] = {
+				mem[0]->getDeviceptr(),
+				mem[1]->getDeviceptr(),
+				mem[2]->getDeviceptr(),
+			};
+			void* parameters[] = { &ptrs[0], &ptrs[1], &ptrs[2], &numElements };
+			ASSERT_SUCCESS(cu.pcuMemcpyHtoDAsync_v2(ptrs[0], cpubuffers[0]->getPointer(), size, stream));
+			ASSERT_SUCCESS(cu.pcuMemcpyHtoDAsync_v2(ptrs[1], cpubuffers[1]->getPointer(), size, stream));
+			ASSERT_SUCCESS(cu.pcuLaunchKernel(kernel, gridDim[0], gridDim[1], gridDim[2], blockDim[0], blockDim[1], blockDim[2], 0, stream, parameters, nullptr));
+			CUDA_EXTERNAL_SEMAPHORE_SIGNAL_PARAMS signalParams = { .params = {.fence = {.value = 1 } } };
+			auto semaphore = cusema->getInternalObject();
+			ASSERT_SUCCESS(cu.pcuSignalExternalSemaphoresAsync(&semaphore, &signalParams, 1, stream)); // Signal the imported semaphore
+
+			std::string abc = "123";
+		}
+		
+		// After the cuda kernel has signalled our exported vk semaphore, we will download the results through the buffer imported from CUDA
+		{
+			IGPUCommandBuffer::SBufferMemoryBarrier<IGPUCommandBuffer::SOwnershipTransferBarrier> bufBarrier = {
+				.barrier = {
+					.dep = {
+						// .srcStageMask = PIPELINE_STAGE_FLAGS::ALL_COMMANDS_BITS,
+						// .srcAccessMask = ACCESS_FLAGS::MEMORY_READ_BITS | ACCESS_FLAGS::MEMORY_WRITE_BITS,
+						.dstStageMask = PIPELINE_STAGE_FLAGS::ALL_COMMANDS_BITS,
+						.dstAccessMask = ACCESS_FLAGS::MEMORY_READ_BITS | ACCESS_FLAGS::MEMORY_WRITE_BITS,
+					},
+					.ownershipOp = IGPUCommandBuffer::SOwnershipTransferBarrier::OWNERSHIP_OP::ACQUIRE,
+					.otherQueueFamilyIndex = queue->getFamilyIndex(),
```

Comment:
no `otherQueueFamilyIndex` is "other" relative to the queue the pipeline barrier (and the commandbuffer it was recorded into) executes on.

so for you its FamilyQueue::External or Foreign here

Reply:
As per https://github.com/Devsh-Graphics-Programming/Nabla-Examples-and-Tests/pull/49#discussion_r1264382481 it should be External not Foreign

Reply:
still stands